### PR TITLE
Add screen sharing support

### PR DESCRIPTION
**Pull request by lizadaily**
This PR adds screen share support to the spatialization demo. Screen sharing can be enabled in the following circumstances:

* A room configuration has screen sharing enabled and one of the following is true:
  * A user is at a desk and another user is not already screen sharing
  * A user is in the broadcast spot

In this way, each user will only be subscribed to a maximum of two screen share tracks at one time. The screen share tile is also draggable .In the below screenshot, the command prompt is being shared in the demo:

image

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
     <meta
       http-equiv="Content-Security-Policy"
-      content="default-src 'self'; frame-src 'self' https://*.daily.co; script-src 'self' https://unpkg.com/@daily-co/daily-js 'unsafe-eval'; connect-src https://*.daily.co https://*.pluot.blue wss:; worker-src 'self' blob:; style-src 'self' https://fonts.googleapis.com/css2; font-src 'self' https://fonts.gstatic.com/s/roboto/v29/KFOmCnqEu92Fr1Mu72xKKTU1Kvnz.woff2 https://fonts.gstatic.com/s/roboto/v29/KFOmCnqEu92Fr1Mu4mxKKTU1Kg.woff2 https://*.gstatic.com"
```
Comment (lizadaily) [inline index.html:10]: This is not related to screen sharing, just removing an unneeded directive

File: style.css [hunk 2]
```diff
@@ -317,12 +345,13 @@ body {
   line-height: 16px;
   color: #ffffff;
   position: absolute;
-  top: 140px;
+  top: calc(100% - 20px);
```
Comment (lizadaily) [inline style.css:348]: We use a calc now because screen share tiles are of a different size than camera tiles, so the height is no longer consistent.

File: src/models/userMedia.ts [hunk 3]
```diff
@@ -405,4 +435,24 @@ export class UserMedia {
     }
     showBroadcast(this.userName, videoTrack, this.audioTrack);
   }
+
+  tryShowScreenShare() {
+    if (!this.screenTrack) return;
+
```
Comment (jessmitch42) [inline src/models/userMedia.ts:441]: Haven't looked through the code enough to know exactly where this comment should go but: you may want to also use the `supportedBrowser` method to see if `supportsScreenShare` is true before trying to screen share too. 

Another thing to consider: you can turn off the screen share option for rooms. You may want to check screen sharing is enabled before actually displaying the screen share button. (Right now, if you try to screen share with a room where it's not enabled, you get a console error but no UI feedback to know what the issue is).

Small thing but would be nice to have :)

Comment (jessmitch42) [reply src/models/userMedia.ts:441]: Ah so I just saw you're already doing this second check mentioned above! but.. something funky is happening. If I join a room where screen sharing is enabled and then leave via the leave button, and then immediately join a room where the screen sharing is not enable, the button still appears. But if I start with the room where it's not enabled, the button doesn't show.

So it looks like the bug is somewhere in resetting things on leave (I would guess).

File: src/room.ts [hunk 4]
```diff
@@ -184,6 +192,25 @@ export class Room {
...
 
+    // Retrieve our room properties and check if screen sharing is enabled
+    const roomProperties = this.callObject
+      .room({ includeRoomConfigDefaults: true })
+      .then((roomInfo) => {
+        const info = roomInfo as DailyRoomInfo;
+        // If screen sharing is disabled, early out
+        if (!info || !info.config?.enable_screenshare) return;
+
+        // If screen sharing is enabled, enable our screen share controls
```
Comment (jessmitch42) [inline src/room.ts:203]: nice :)

File: src/room.ts [hunk 5]
```diff
@@ -184,6 +192,25 @@ export class Room {
     // Get the local participant
     const p = event.participants["local"];
 
+    // Retrieve our room properties and check if screen sharing is enabled
+    const roomProperties = this.callObject
+      .room({ includeRoomConfigDefaults: true })
+      .then((roomInfo) => {
+        const info = roomInfo as DailyRoomInfo;
```
Comment (jessmitch42) [inline src/room.ts:199]: does this need to be reassigned? (I'm guessing it's a typescript thing..)

File: src/models/userMedia.ts (hunk 3 above)
Comment (lizadaily) [reply src/models/userMedia.ts:441]: Ah ha good to know, I'll fix that tomorrow (and add the `supportedBrowser` check as well), thanks!

File: src/room.ts [hunk 6]
```diff
@@ -374,4 +417,28 @@ export class Room {
...
   }
+
+  private async startScreenShare() {
+    let captureStream = null;
+    const options = {
+      video: true,
+    };
+
+    try {
+      captureStream = await navigator.mediaDevices.getDisplayMedia(options);
```
Comment (jessmitch42) [inline src/room.ts:428]: what does this do? I usually just call startScreenShare so just curious

**Review by jessmitch42 — COMMENTED** [review]
This all looks good to me! Added one suggested and one little bug I noticed in the comments

File: src/room.ts (hunk 6 above)
Comment (lizadaily) [reply src/room.ts:428]: This allows us to grab the MediaStream manually with our own constraints, which Dominic mentioned [on Slack](https://pluot.slack.com/archives/C01M0U5TZA6/p1646408198877319?thread_ts=1645204505.791569&cid=C01M0U5TZA6) is a common use case with spatialization customers, so we figured it might be good to showcase how to do that here.

Comment (jessmitch42) [reply src/room.ts:428]: very cool, thanks!

File: src/room.ts (hunk 5 above)
Comment (lizadaily) [reply src/room.ts:199]: Afraid so; I first tried to cast this to the right type in the callback parameter itself but that didn't work (will look some more tomorrow just in case I missed another way)

File: index.html [hunk 7]
```diff
@@ -72,15 +72,21 @@ <h1 class="title">Spatialization demo</h1>
         <div id="controls">
           <button id="toggleCam"></button>
           <button id="toggleMic"></button>
+          <button
+            id="toggleScreenShare"
+            class="screen-disabled hidden"
+            disabled
+          ></button>
```
Comment (jayne-mast) [inline index.html:79]: Accessibility-wise, it would be better if these buttons had content in it that describes the action.

Comment (lizadaily) [reply index.html:79]: Good point - would this be primarily for screen readers? (ie, would it work to set the text in the code but then have it be invisible in the UI itself?) Thanks for pointing this out.

File: src/util/tile.ts [hunk 8]
```diff
@@ -26,79 +35,141 @@ export function showBroadcast(
...
+}
+
+// showScreenShare shows a screen share video tile in the zonemates div
+export function showScreenShare(
+  sessionID: string,
+  name: string,
+  videoTrack?: MediaStreamTrack
+) {
+  const tracks: Array<MediaStreamTrack> = [];
+  if (videoTrack) tracks.push(videoTrack);
```
Comment (jayne-mast) [inline src/util/tile.ts:63]: I haven't seen us using single-line if-statements before,

Comment (lizadaily) [reply src/util/tile.ts:63]: I _think_ I've seen this in other demos and also in pluot-core itself (but I could be wrong). For me this just looks more readable here as we are doing two nearly identical push operations right next to each other. Is there a reason not to use them?

File: index.html (hunk 7 above)
Comment (jayne-mast) [reply index.html:79]: Yes, it's just for screen readers. We could use `aria-label` instead of content, so we don't have to hide it in a weird CSS way.

Comment (lizadaily) [reply index.html]: Done!

**Review by jayne-mast — APPROVED** [review]

File: src/util/tile.ts (hunk 8 above)
Comment (jessmitch42) [reply src/util/tile.ts:63]: The rule of thumb I've used for if statements is a single line if it's just a return and braces if there's any action. 

```
if (whatever) return;
```
We don't have this sort of thing in the style guide as of yet so it's pretty subjective.

**Comment by jessmitch42** [comment]
@lizadaily one little edge case I found: If I leave the broadcast zone while screen sharing, it ends the screen share but doesn't disable the screen share button.
CleanShot 2022-03-14 at 10 53 06

**Comment by lizadaily** [comment]
> @lizadaily one little edge case I found: If I leave the broadcast zone while screen sharing, it ends the screen share but doesn't disable the screen share button. CleanShot 2022-03-14 at 10 53 06

Thanks, great catch! This was related to a timing issue, where Daily's `"participant-updated"` event with no screen track fired _after_ we tried to disable the button, so it got caught on a safety check where we tried to ensure we don't disable the button while the user is still screen sharing. I think it's reasonable to remove this check because we have direct control of the track here and don't depend on Daily stopping it for us, so no need for us to wait for Daily's event. Should be fixed now!

**Review by jessmitch42 — APPROVED** [review]
lgtm 🎉
